<model>: clip entities on the page z level

amarchenko opened the pull request:
#### 59d90a1043ebf2c49399cc1421ed06b5e729cc7b
<pre>
&lt;model&gt;: clip entities on the page z level
<a href="https://rdar.apple.com/125188888">rdar://125188888</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=272183">https://bugs.webkit.org/show_bug.cgi?id=272183</a>

Reviewed by Tim Horton.

Initial implementation of clipping on +Z direction should use PortalComponent&apos;s
ClippingPlane but it is blocked by <a href="https://rdar.apple.com/123516357">rdar://123516357</a>.
Current implementation uses a workaround based on ClippingPrimitive. Bounding box
of the primitive is an arbitrary number that is big enoungh for the model representation.

* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::didFinishLoading):

Canonical link: <a href="https://commits.webkit.org/277138@main">https://commits.webkit.org/277138@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8680060e97325cce987a38e2e33abb216aebdf36

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46756 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25917 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49378 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49433 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42803 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/30302 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23380 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/38100 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47336 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22913 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40269 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19386 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/20252 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4802 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43006 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41769 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51305 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21765 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18123 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45380 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23053 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44337 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10340 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/23542 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22759 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -352,11 +350,31 @@ static CGFloat effectivePointsPerMeter(CALayer *caLayer)
-    RENetworkMarkEntityMetadataDirty(rootEntity);
+
+    // FIXME: Clipping workaround for rdar://125188888 (blocked by rdar://123516357 -> rdar://124718417).
+    // middleEntity is required to add a clipping primitive that is independend from model's rootEntity.
```
Nit
```suggestion
    // middleEntity is required to add a clipping primitive that is independent from model's rootEntity.
```

amarchenko replied:
Nice catch. Fixed.

eddydas commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -352,11 +350,31 @@ static CGFloat effectivePointsPerMeter(CALayer *caLayer)
-    auto rootEntity = REComponentGetEntity(clientComponent);
-    REEntitySetName(rootEntity, "WebKit:ClientComponentRoot");
+    auto clientComponentEntity = REComponentGetEntity(clientComponent);
+    REEntitySetName(clientComponentEntity, "WebKit:ClientComponentRoot");
```
Let's also call it "WebKit:ClientComponentEntity"

amarchenko replied:
Done.

eddydas commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -352,11 +350,31 @@ static CGFloat effectivePointsPerMeter(CALayer *caLayer)
+    // middleEntity is required to add a clipping primitive that is independent from model's rootEntity.
+    // Adding the primitive directly to clientComponentEntity has no visual effect.
+    constexpr float clippingBoxHalfSize = 500; // meters
+    REPtr<REEntityRef> middleEntity = adoptRE(REEntityCreate());
```
I'm fine with "middle" since I think this should go away once we move to clipping plane implementation.
But if this is to stay, I'd think "containerEntity" may be a better name than "middleEntity"

amarchenko replied:
Agree. Changed to `containerEntity`.

eddydas commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -352,11 +350,31 @@ static CGFloat effectivePointsPerMeter(CALayer *caLayer)
+    REEntitySubtreeAddNetworkComponentRecursive(middleEntity.get());
+
+    auto clipComponent = REEntityGetOrAddComponentByClass(middleEntity.get(), REClippingPrimitiveComponentGetComponentType());
+    REClippingPrimitiveComponentSetShouldClipChildren(clipComponent, false);
```
Do you mean "true" here?

amarchenko replied:
It must be `true`. Fixed .

eddydas reviewed (commented):
LGTM.

amarchenko committed:
<model>: clip entities on the page z level
rdar://125188888
https://bugs.webkit.org/show_bug.cgi?id=272183

Reviewed by Tim Horton.

Initial implementation of clipping on +Z direction should use PortalComponent's
ClippingPlane but it is blocked by rdar://123516357.
Current implementation uses a workaround based on ClippingPrimitive. Bounding box
of the primitive is an arbitrary number that is big enoungh for the model representation.

* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::didFinishLoading):

Canonical link: https://commits.webkit.org/277138@main
